This pull request improves the robustness of metadata parsing for XTTS fine-tuning and fixes a critical bug.

porhe911 opened the pull request:
Fixed: AttributeError caused by attempting to call .strip() on None values when parsing CSV/metadata files.
Improved: Metadata parsing now safely handles None values—both keys and values in CSV files are converted to empty strings if missing.
Benefit: Prevents crashes during dataset preparation when metadata files contain incomplete or invalid data.
Result: More stable and reliable data processing workflow for XTTS fine-tuning in TTS-WebUI.

Due to this error, I could not train the model in TTS WebUI (XTTS Fine-tuning Demo).Because of this, neither the 1st tab (data processing) nor the second (fine-tuning xtts encoding) worked. I fixed this issue in main.py because that was the problem. If you need screenshots or details, I am ready to provide!
And special thanks to you for TTS WebUI, I like this project and am ready to help you with it!
<img width="1918" height="1004" alt="Screenshot_4" src="https://github.com/user-attachments/assets/a56826e9-4899-45ce-9e0e-5287716e3a6c" />
<img width="1919" height="1008" alt="Screenshot_1" src="https://github.com/user-attachments/assets/440da868-cce8-432a-aa22-3f76f5f3e746" />

porhe911 committed:
Fix dataset normalization and progress handling

porhe911 committed:
Merge pull request #1 from porhe911/codex/fix-errors-in-tts-webui-training

Fix progress adapter and metadata normalization for XTTS fine-tuning

porhe911 committed:
fix: robust handling of None values in metadata CSV parsing

- Fixed AttributeError: 'NoneType' object has no attribute 'strip' when parsing CSV/metadata files.
- Metadata parsing now safely handles None values: CSV keys and values that are None are converted to empty strings.
- This improves stability when loading and processing audio datasets with incomplete or invalid metadata.
- Data preparation for XTTS fine-tuning is now more reliable and less prone to crashes.
------------------------------------------------------------------------
- Исправлена ошибка AttributeError: 'NoneType' object has no attribute 'strip' при парсинге CSV/метаданных.
- Теперь обработка строк из метаданных безопасно работает с None: ключи и значения, равные None, преобразуются в пустые строки.
- Благодаря этому, при загрузке и обработке аудиоданных с некорректными или неполными метаданными не возникает сбоев.
- Улучшена стабильность и надёжность пайплайна подготовки данных для XTTS fine-tuning.

porhe911 committed:
fix: auto delimiter correction for train/eval and pydantic progress integer bug

- Added automatic conversion of train.csv and eval.csv from comma-separated to pipe-separated format (|) to avoid parsing errors and ensure XTTS compatibility.
- During data preprocessing and before training, train and eval files are now automatically converted to the required format, even if originally comma-separated.
- Fixed pydantic "Input should be a valid integer, got a number with a fractional part" error — total for tqdm/gradio progress is always cast to an integer.
- Improved pipeline stability and UX: no more errors due to incorrect dataset format or progress tracking.
------------------------------------------------------------------------
- Добавлена функция автоматического преобразования train.csv и eval.csv из формата с запятыми в формат с вертикальной чертой (|) — чтобы избежать ошибок парсинга и обеспечить совместимость с XTTS.
- Теперь при препроцессинге данных и перед запуском обучения train и eval автоматически конвертируются в нужный формат, даже если исходно были разделены запятыми.
- Исправлена ошибка pydantic "Input should be a valid integer, got a number with a fractional part" — теперь total для tqdm/gradio всегда приводится к целому числу.
- Улучшена стабильность пайплайна и UX: больше не возникнут ошибки из-за неправильного формата датасета или прогресса.

rsxdalv commented:
Was there any issue that caused you to close this PR?
I reviewed half of it, and wasn't sure about the class made for compatibility but otherwise looked good

porhe911 commented:
Good afternoon, yes, I found the problem in this code myself, I will fix it today and tomorrow I will offer fixes to you. I'm testing everything before shipping. I had an error in the code related to correcting the csv with each training cycle, because of this, the original text data was distorted. I use: Windows 10 Pro
Version 22H2
Installation date 07/27/2024
OS Build 19045.6332
But my code solves the error in 1 window.

porhe911 committed:
This update improves training stability and dataset preprocessing for XTTS fine-tuning.

feat(csv): add automatic safe conversion of train/eval CSV to pipe format
fix(tqdm): ensure total is always cast to int for proper progress display
refactor(io): replace os.system("cp") with shutil.copy for cross-platform safety
------------------------------------------------------------------------feat(csv): автоматическая безопасная конвертация train/eval CSV в pipe-формат
fix(tqdm): принудительное преобразование total в int для корректного прогресса
refactor(io): замена os.system("cp") на shutil.copy для кроссплатформенности

porhe911 commented:
It seems that it turned out to fix everything, how will you have time to test it yourself? My initial problem was due to the fact that the csv files had a split "," and needed |
Now this is done automatically, despite the fact that the commas in the text do not change to | but if you already have separators | then there will be no replacement.

This PR improves the XTTS fine-tuning process by adding **automatic CSV normalization**, a **tqdm progress fix**, and **cross-platform file copying**.

It ensures that `train.csv` and `eval.csv` are always pipe-delimited (`|`) for XTTS training while preserving full text integrity and compatibility with Windows and Linux.

Changes:
- Added `fix_csv_delimiters(filepath)`:
- Automatically detects the delimiter using `csv.Sniffer`.
- Converts comma-delimited CSVs to pipe (`|`) **only when needed**.
- Preserves commas and quotes inside text fields — no truncation.
- Uses `newline=""` to avoid extra blank lines on Windows.
- Pads rows to at least 4 fields (`wav_path|speaker|language|text`).
- Leaves already pipe-formatted files untouched.
- Invoked after preprocessing and before training for consistency.
- Prevents progress-bar freezes or crashes when `total` is a float or numpy value.
- Replaced `os.system("cp ...")` with `shutil.copy(...)`.
- Works natively on both Windows and Linux — no dependency on external `cp`.

rsxdalv commented on tts_webui_extension/xtts_ft_demo/main.py:
```diff
@@ -209,8 +236,11 @@ def preprocess_dataset(
-                        gradio_progress=progress,
+                        gradio_progress=None,
                     )
+                    # Автоматически фиксируем формат после препроцессинга
```
```suggestion
                    # Automatically fix the format after preprocessing
```

Keep comments in  English

rsxdalv commented on tts_webui_extension/xtts_ft_demo/main.py:
```diff
@@ -298,7 +293,9 @@ def train_model(
                 )
             try:
-                # convert seconds to waveform frames
+                # Автоматически фиксируем формат перед обучением
```
```suggestion
                # Automatically fix the format before training
```

rsxdalv committed:
Update tts_webui_extension/xtts_ft_demo/main.py

rsxdalv committed:
Update tts_webui_extension/xtts_ft_demo/main.py

rsxdalv commented:
Thank you!
